src/block.js

```diff
@@ -0,0 +1,201 @@
+        for (let i = 0; i < val.length; i++) {
+          const __path = _path.slice()
+          __path.push(i)
+          const o = val[i]
+          yield __path.join('/')
+          if (typeof o === 'object' && !CID.asCID(o)) {
+            yield * tree(o, __path)
+          }
+        }
+      } else {
+        yield * tree(val, _path)
+      }
+    }
+  }
+}
+/* eslint-enable max-depth */
+
+class Block {
+  constructor ({ codec, hasher, source, cid, data }) {
+    setImmutable(this, '_codec', codec)
+    setImmutable(this, '_hasher', hasher)
+    if (source) setImmutable(this, '_source', source)
+    if (cid) setImmutable(this, '_cid', cid)
+    if (data) setImmutable(this, '_data', data)
+    setImmutable(this, 'asBlock', this)
+  }
+
+  decodeUnsafe () {
+    if (typeof this._source !== 'undefined') return this._source
+    throw new Error('Block created without a decoded state')
+  }
+
+  decode () {
+    const decoded = this.decodeUnsafe()
+    if (decoded === null) return null
+    if (isBinary(decoded)) return copyBinary(decoded)
+    if (immutableTypes.has(typeof decoded) || decoded === null) {
+      return decoded
+    }
+    return clone(decoded)
+  }
+
+  encodeUnsafe () {
+    if (this._data) return this._data
+    throw new Error('Block created without an encoded state')
+  }
+
+  encode () {
+    const data = this.encodeUnsafe()
+    return copyBinary(data)
+  }
+
+  async cid () {
+    if (this._cid) return this._cid
+    const hash = await this._hasher.digest(this.encodeUnsafe())
+    const cid = CID.create(1, this._codec.code, hash)
+    setImmutable(this, '_cid', cid)
+    return cid
+  }
+
+  get code () {
+    if (this._cid) return this._cid.code
+    return this._codec.code
+  }
+
+  async equals (block) {
+    if (block === this) return true
+    if (block.asBlock !== block) return false
+    const [a, b] = await Promise.all([this.cid(), block.cid()])
+    return a.equals(b)
+  }
+
+  links () {
+    return links(this.decodeUnsafe())
+  }
+
+  tree () {
+    return tree(this.decodeUnsafe())
+  }
+}
+
+class BlockDecoder extends Block {
+  decodeUnsafe () {
+    if (typeof this._source !== 'undefined') return this._source
+    if (!this._codec) {
+      throw new Error('Do not have codec implemention in this Block interface')
+    }
+    const source = this._codec.decode(this._data)
+    setImmutable(this, '_source', source)
+    return source
+  }
+}
+class BlockEncoder extends Block {
+  encodeUnsafe () {
+    if (this._data) return this._data
+    if (!this._codec) {
+      throw new Error('Do not have codec implemention in this Block interface')
+    }
+    const data = this._codec.encode(this._source)
+    setImmutable(this, '_data', data)
+    return data
+  }
+}
+
+const encoder = ({ source, codec, hasher }) => {
+  if (typeof source === undefined) throw new Error('Missing required argument "source"')
+  if (!codec || !hasher) throw new Error('Missing required argument: codec or hasher')
```

Comment:
I think it would be better to have separate errors for different options so it's easier to figure out the cause.

```suggestion
  if (!codec) throw new Error('Block decoder option ".codec" must be provided')
  if (!hasher) throw new Error('Option ".hasher" must be provided')
```